File: core/src/main/java/xtdb/vector/IVectorPosition.java
package xtdb.vector;

public interface IVectorPosition {
    int getPosition();
    void setPosition(int position);

    default int getPositionAndIncrement() {
        int position = getPosition();
        this.setPosition(position + 1);
        return position;
    }

    static IVectorPosition build() {
        return build(0);
    }

    static IVectorPosition build(int initialPosition) {
        return new IVectorPosition() {
            private int position = initialPosition;

            @Override
            public int getPosition() {
                return position;
            }

            @Override
            public void setPosition(int position) {
                this.position = position;
            }
        };
    }
}

File: core/src/main/kotlin/xtdb/vector/IVectorPosition.kt
package xtdb.vector

interface IVectorPosition {
    var position: Int

    fun getPositionAndIncrement() = position++

    companion object {
        @JvmOverloads
        @JvmStatic
        fun build(initialPosition: Int = 0): IVectorPosition {
            return object : IVectorPosition {
                override var position: Int = initialPosition
            }
        }
    }
}
